uhc: dwc2: vendor quirks for nRF54LM20

josuah opened the pull request:
- https://github.com/zephyrproject-rtos/zephyr/issues/102931

Dependencies:

- https://github.com/zephyrproject-rtos/zephyr/pull/102967

For testing, provide power externally using an USB OTG adapter with external power supply.

josuah commented:
So far, only getting the IRQs upon connection/disconnection are working, and almost none of the content of `uhc_dwc2.c` is present.

roma-jam commented:
Hi @josuah, 

I have removed the LL-layer, so please, feel free to rebase. 

Also, I can see that the log is from uhc and then in the end - the udc: `uhc_dwc2_enable: Enable device 0x50020000`. 

Just in case - is that intentional? Because I didn't check the possibility to run udc after uhc, so there could be some moments, that need to be fixed.

josuah commented:
Some script showing the difference in hardware configuration registers of ESP32-S3 and nRF54LM20.

[`dwc2_decode_regs.py`](https://github.com/user-attachments/files/23401652/dwc2_decode_regs.py)

<details>
<summary>output</summary>

    
    GHWCFG1 (ESP32): 0x00000000
     EpDir:32 = 0
    32 bits
    
    GHWCFG1 (nRF54): 0x00000000
     EpDir:32 = 0
    32 bits
    
    GHWCFG2 (ESP32): 0x224dd930
     OtgMode:3 = 0
     OtgArch:2 = 2
     SingPnt:1 = 1
     HsPhyType:2 = 0
     FsPhyType:2 = 1
     NumDevEps:4 = 6
     NumHstChnl:4 = 7
     PerioSupport:1 = 1
     DynFifoSizing:1 = 1
     MultiProcIntrpt:1 = 0
     Reserved:1 = 0
     NPTxQDepth:2 = 1
     PTxQDepth:2 = 2
     TknQDepth:5 = 8
     OTG_ENABLE_IC_USB:1 = 0
    32 bits
    
    GHWCFG2 (nRF54): 0x22affc52
     OtgMode:3 = 2
     OtgArch:2 = 2
     SingPnt:1 = 0
     HsPhyType:2 = 1
     FsPhyType:2 = 0
     NumDevEps:4 = 15
     NumHstChnl:4 = 15
     PerioSupport:1 = 1
     DynFifoSizing:1 = 1
     MultiProcIntrpt:1 = 0
     Reserved:1 = 1
     NPTxQDepth:2 = 2
     PTxQDepth:2 = 2
     TknQDepth:5 = 8
     OTG_ENABLE_IC_USB:1 = 0
    32 bits
    
    GHWCFG3 (ESP32): 0x00c804b5
     XferSizeWidth:4 = 5
     PktSizeWidth:3 = 3
     OtgEn:1 = 1
     I2CIntSel:1 = 0
     VndctlSupt:1 = 0
     OptFeature:1 = 1
     RstType:1 = 0
     ADPSupport:1 = 0
     HSICMode:1 = 0
     BCSupport:1 = 0
     LPMMode:1 = 0
     DfifoDepth:16 = 200
    32 bits
    
    GHWCFG3 (nRF54): 0x0be0c0e8
     XferSizeWidth:4 = 8
     PktSizeWidth:3 = 6
     OtgEn:1 = 1
     I2CIntSel:1 = 0
     VndctlSupt:1 = 0
     OptFeature:1 = 0
     RstType:1 = 0
     ADPSupport:1 = 0
     HSICMode:1 = 0
     BCSupport:1 = 1
     LPMMode:1 = 1
     DfifoDepth:16 = 3040
    32 bits
    
    GHWCFG4 (ESP32): 0xd3f0a030
     NumDevPerioEps:4 = 0
     PartialPwrDn:1 = 1
     AhbFreq:1 = 1
     Hibernation:1 = 0
     ExtendedHibernation:1 = 0
     Reserved_8:1 = 0
     EnhancedLPMSupt1:1 = 0
     ServIntFlow:1 = 0
     ipgisocSupt:1 = 0
     ACGSupt:1 = 0
     EnhancedLPMSupt:1 = 1
     PhyDataWidth:2 = 2
     NumCtlEps:4 = 0
     IddgFltr:1 = 1
     VBusValidFltr:1 = 1
     AValidFltr:1 = 1
     BValidFltr:1 = 1
     SessEndFltr:1 = 1
     DedFifoMode:1 = 1
     INEps:4 = 4
     DescDMAEnabled:1 = 1
     DescDMA:1 = 1
    32 bits
    
    GHWCFG4 (nRF54): 0x3e10aa60
     NumDevPerioEps:4 = 0
     PartialPwrDn:1 = 0
     AhbFreq:1 = 1
     Hibernation:1 = 1
     ExtendedHibernation:1 = 0
     Reserved_8:1 = 0
     EnhancedLPMSupt1:1 = 1
     ServIntFlow:1 = 0
     ipgisocSupt:1 = 1
     ACGSupt:1 = 0
     EnhancedLPMSupt:1 = 1
     PhyDataWidth:2 = 2
     NumCtlEps:4 = 0
     IddgFltr:1 = 1
     VBusValidFltr:1 = 0
     AValidFltr:1 = 0
     BValidFltr:1 = 0
     SessEndFltr:1 = 0
     DedFifoMode:1 = 1
     INEps:4 = 15
     DescDMAEnabled:1 = 0
     DescDMA:1 = 0
    32 bits
    
    GOTGCTL (ESP32): 0x00240000
     SesReqScs:1 = 0
     SesReq:1 = 0
     VBValidOvEn:1 = 0
     VBValidOvVal:1 = 0
     AValidOvEn:1 = 0
     AValidOvVal:1 = 0
     BValidOvEn:1 = 0
     BValidOvVal:1 = 0
     HstNegScs:1 = 0
     HnpReq:1 = 0
     HstSetHnpEn:1 = 0
     DevHnpEn:1 = 0
     EhEn:1 = 0
     RSVD:2 = 0
     DbnceFltrBypass:1 = 0
     ConIdSts:1 = 0
     DbncTime:1 = 0
     ASesVld:1 = 1
     BSesVld:1 = 0
     OtgVer:1 = 0
     CurMod:1 = 1
     MultValidBc:5 = 0
     ChirpEn:1 = 0
     eUSB2_Phy_Disc_Supp:1 = 0
     Reserved_30_29:2 = 0
     TestMode_Corr_eUSB2:1 = 0
    32 bits
    
    GOTGCTL (nRF54): 0x03050070
     SesReqScs:1 = 0
     SesReq:1 = 0
     VBValidOvEn:1 = 0
     VBValidOvVal:1 = 0
     AValidOvEn:1 = 1
     AValidOvVal:1 = 1
     BValidOvEn:1 = 1
     BValidOvVal:1 = 0
     HstNegScs:1 = 0
     HnpReq:1 = 0
     HstSetHnpEn:1 = 0
     DevHnpEn:1 = 0
     EhEn:1 = 0
     RSVD:2 = 0
     DbnceFltrBypass:1 = 0
     ConIdSts:1 = 1
     DbncTime:1 = 0
     ASesVld:1 = 1
     BSesVld:1 = 0
     OtgVer:1 = 0
     CurMod:1 = 0
     MultValidBc:5 = 12
     ChirpEn:1 = 0
     eUSB2_Phy_Disc_Supp:1 = 0
     Reserved_30_29:2 = 0
     TestMode_Corr_eUSB2:1 = 0
    32 bits
    
    HCFG (nRF54): 0x00220000
     FSLSPclkSel:2 = 0
     FSLSSupp:1 = 0
     RESERVED2:4 = 0
     Ena32KHzS:1 = 0
     ResValid:8 = 0
     dis_tx_ipgap_dly_check:1 = 0
     RESERVED1:6 = 17
     DescDMA:1 = 0
     FrListEn:2 = 0
     PerSchedEna:1 = 0
     RESERVED:4 = 0
     ModeChTimEn:1 = 0
    32 bits
    
    HCFG (ESP32): 0x00100000
     FSLSPclkSel:2 = 0
     FSLSSupp:1 = 0
     RESERVED2:4 = 0
     Ena32KHzS:1 = 0
     ResValid:8 = 0
     dis_tx_ipgap_dly_check:1 = 0
     RESERVED1:6 = 8
     DescDMA:1 = 0
     FrListEn:2 = 0
     PerSchedEna:1 = 0
     RESERVED:4 = 0
     ModeChTimEn:1 = 0
    32 bits
    
    HFIR (nRF54): 0x000017d7
     FrInt:16 = 6103
     HFIRRldCtrl:1 = 0
     RESERVED:17 = 0
    34 bits
    
    HFIR (ESP32): 0x000017d7
     FrInt:16 = 6103
     HFIRRldCtrl:1 = 0
     RESERVED:17 = 0
    34 bits


</details>

josuah commented:
2a120d052025908200d0751ce7ea7d6c766ce2a9 contains ~100 bissectable mini-commits that slowly transform the original work passed from hand to hand into a hopefully more minimal and workable following a single coding/naming style and uniform structure (`git diff --stat 830c8b5ce71` says `1005 insertions(+), 1922 deletions(-)`).

roma-jam commented:
@josuah  wow, I briefly check the changes....looks more Zephyr-style :)

I also briefly checked the most critical parts and seems that that logic was kept (at least from the first glance).
I'll try to check it on a hardware asap. finish with other tasks and get back to this one.

josuah commented:
Thank you.

> Ok, back to the topic: would you mind if I cherry-pick your commits with refactoring to my branch and:

Please go ahead, this was the original intent of this PR: to let everyone working on DWC2 host keep going on a common ground.

> - verify one more time the logic and check it on a hardware

I observe there is some problem while starting Zephyr and USB init/enable first, and plugging the USB device after.

I can use the "[bisectable commits](https://github.com/zephyrproject-rtos/zephyr/pull/95723#issuecomment-3524633466)" above to check which small change broke something if that helps.

> - proceed with the rest TODOs

I will have to switch to another project soon, hopefully this helps getting this branch stable as you start using it.

I will keep this one as draft just for adding the nRF54LM20 and will rebase it on top of your changes as you publish them.

roma-jam commented:
@josuah , 

> I observe there is some problem while starting Zephyr and USB init/enable first, and plugging the USB device after.

I remember I already had the same problem two times during the past time and both times I solved it.
I  will take a look.

josuah commented:
Hi @roma-jam,

> I think that it might be better to keep the correct serialization of events from the port. So it will be a bit easier to handle them in task.

I remember there was several layers of events, so it was decoding of decoding of events, or something like this.

Maybe this is possible to go back to the pseudocode from the *OTG programming guide 5.1.2.3 (p. 139)*? Or were you referring to some particular code construct that I removed in the process of merging things? I remember there were indications of which order things were to be executed in the comments.

> Now I can see that I am able to use the device address from the `xfer->udev->addr` and configure the channel by request.

👍

Regarding the allocation of channels' FIFO sizes, there is this PR to allow to statically allocate it, and simplify the logic:

- https://github.com/zephyrproject-rtos/zephyr/pull/99510

For the actual channels, I took the bet that it would be relatively small amount of data to allocate them all in advance.

> Also, I would like to ask your opinion (because of the amount of the code) which option is better:
> [...]

I have the impression we need to have some driver that can be tested merged in. What would be the minimal version?

Maybe something that just supports the control channel, with most features set as `LOG_ERR("not implemented");`?

> If you think it is better to work separately, this is also fine, So, please let me know.
> PS. I am not tied to Espressif anymore, so I can switch the target to Nordic and proceed with its hardware.

Would it be more fair to consider the contributions best effort?
Regarding nRF54LM20, it seems like it is slowly rolling out, but still difficult to access, but I can always ask.

Either way, it will help to be technically independent (i.e. a common base on top of which features are added) and communicate as often as it helps.

Maybe it is possible to polish the current version and submit it, now that enumerating various devices works.

Thanks for reaching in.

roma-jam commented:
@josuah, 

> Maybe this is possible to go back to the pseudocode from the OTG programming guide 5.1.2.3 (p. 139)?

I don't have access to the OTG programming guide anymore, but I will check what I can do. Sure, it will be better to be as close as possible to the guide. 

> For the actual channels, I took the bet that it would be relatively small amount of data to allocate them all in advance.

Yes, that was my original idea. And not to lock the channel to the EP, but to occupy the vacant channel and configure it on xfer enqueue. That will help us to use channels whether they are free and not to limit the interfaces for the uhc.

> I have the impression we need to have some driver that can be tested merged in. What would be the minimal version?

IMHO, that should be: 
- correct device connect/disconnect handling (the best case scenario: several hundreds attempts in a row with different delay between conn/dconn)
- CTRL XFER logic (set address, get descriptor, ...)

But, we already have the common struct for the basic functions (API, basic quirks and controller description). I can curve out this and create a PR with only this, common part. Please, let me know if that might help, Otherwise, I will proceed with the logic, described just a bit higher. 

> Maybe it is possible to polish the current version and submit it, now that enumerating various devices works.

Yes, at least this will help us to move forward. I checked again the Buffer DMA mode and we can omit the recovery state completely, as it needed only for recovering during the Scatter Gather DMA mode. 

So, I will double check the code again.

roma-jam commented:
Ok, lets start with small steps. 

If it is possible, PTAL: #101886

josuah commented:
> Ok, lets start with small steps.
> If it is possible, PTAL: https://github.com/zephyrproject-rtos/zephyr/pull/101886

I see better what you mean by small changes and that makes sense.

We might need to wait the USB maintainer to be available again, but possible to keep progressing in the meantime of course.

roma-jam commented:
Hi @josuah, 

> Up to reviewers to skip the commits of the previous PRs then?

Without approve and merging - it will be impossible to proceed. So, anyway they need to review and approve. The question is just how big is the PR. 

> I typically open everything on top of main, and state the dependencies in the first issue (like you did).

Sometimes, when they are in dependency, it is not possible to open them. In this case the way is just to aim to the branch, to see the diff correctly. Otherwise, if all of them pointed to the main, the last one will have diff of all the previous. 

> Would you wish me to annotate fully your PR so you can decide which change to import, so that the PR you contain already have the needed changes?

You got the idea, so it is up to you. Anyway, after the https://github.com/zephyrproject-rtos/zephyr/pull/101886 being merged, I will move the https://github.com/roma-jam/zephyr/pull/1 to the upstream, so you will be in reviewers there, anyway. 
 
You will be able to see the small diff, align the changes with your version and if necessary we can change it in advance. 
I don't see the reason to review them in my fork and then one more time in the upstream.

roma-jam commented:
Hi @josuah, 

any update with the host class API and UVC?

Sorry to poke you in here, if there is a better place to ask you such questions, let me know.

josuah commented:
Hello @roma-jam. The topic of USB Host Class arrived in today's meeting and will bring a summary of DWC2 soon on this issue. Thank you for waiting these 3 days.

I have been starting to implement BULK, Isochronous, Interrupt, together because it was simpler WRT documentation.

In parallel, I also started to re-implement the event decoding and am currently debugging it.

In parallel, I also made some light changes to this PR which are not pushed yet.

You may find all of it in the latest commit of this branch:

- https://github.com/josuah/zephyr/tree/dwc2_host_wip

This does not currently work but it does contain all the event decoding according to the datasheet.

I will see if it is possible to fit it in the IRQ (for better reactivity), or if it is required to decode things in the event handler.

Thank you for the progress!

josuah commented:
I just completed the decoding of events following the datasheet.

A deep tree of `if ()` that specifically handle each and every case and give us a set of event that we can then use to act upon the IRQ.

This is expected to handle split transactions as well as control/bulk/isochronous/interrupt but only control are supported end-to-end still.

This branch also now contains a `uhc_dwc2_process_ctrl_xfer()` separate from `uhc_dwc2_process_data_xfer()`, anticipating support for non-control transfers.

This should also help with `STALL` requests in the future, and the driver fails with an error when that happens currently.

@roma-jam do you like these changes? Do you wish to integrate them on your branch? Do I instead integrate your changes in this branch?

Commit:
boards: nrf54lm20dk: enable USB Host support

Enable the DWC2 host controller, for now requiring external power to be
provided to the VBUS pin, however usable.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

Commit:
drivers: usb: uhc: dwc2: add nRF54LM20 vendor quirks

Add vendor quirks for nRF54LM20 PHY configuration for DWC2 in host mode.
Given the PHY is powered from USB, the PHY needs to be configured and
powered on before registers can be accessed or the SoC freezes on any
register access.

Signed-off-by: Josuah Demangeon <josuah.demangeon@nordicsemi.no>

josuah commented:
Force-push:

- Rebase: https://github.com/zephyrproject-rtos/zephyr/pull/102967
- Import: https://github.com/zephyrproject-rtos/zephyr/pull/104359

roma-jam commented on drivers/usb/uhc/uhc_dwc2_nrf_usbhs_nrf54l.h:
```diff
@@ -0,0 +1,197 @@
+	int err;
+
+	/* TODO: not expected to be there */
+	LOG_MODULE_DECLARE(uhc_dwc2, CONFIG_UHC_DRIVER_LOG_LEVEL);
```
might be removed now

josuah replied:
Thank you.

roma-jam commented on drivers/usb/uhc/uhc_dwc2_nrf_usbhs_nrf54l.h:
```diff
@@ -0,0 +1,197 @@
+
+	if (!k_event_wait(&data->events, USBHS_VBUS_READY, false, K_NO_WAIT)) {
+		LOG_INF("VBUS is not ready, block udc_enable()");
+		if (!k_event_wait(&data->events, USBHS_VBUS_READY, false, timeout)) {
```
Why do we need to wait USBHS_VBUS_READY two times: one with `K_NO_WAIT` and he second one with `timeout`?

Isn't the `timeout` only enough?

jfischer-no replied:
That is from the UDC DWC2 quirk, the first one does not wait but allows to print a logging message. That does not seems to be right for the UHC.

josuah replied:
Done, only one `k_event_wait()` now.

The message was printed every time in practice...

josuah commented:
Force-push:

- Small bugfix for DWC2 to support nRF54LM20

Not sure why this is needed, but it helps to do this:

```diff
diff --git a/drivers/usb/uhc/uhc_dwc2.c b/drivers/usb/uhc/uhc_dwc2.c
index d4453c20cdd..c249674e5e9 100644
--- a/drivers/usb/uhc/uhc_dwc2.c
+++ b/drivers/usb/uhc/uhc_dwc2.c
@@ -847,7 +847,6 @@ static int uhc_dwc2_channel_configure(const struct device *const dev,
 	hcintmsk = sys_read32((mem_addr_t)&channel->regs->hcintmsk);
 
 	/* Enable transfer complete and channel halted interrupts */
-	hcintmsk |= USB_DWC2_HCINT_XFERCOMPL;
 	hcintmsk |= USB_DWC2_HCINT_CHHLTD;
 	/* Enable error interrupts */
 	hcintmsk |= USB_DWC2_HCINT_ERRORS;
@@ -1304,6 +1303,8 @@ static int uhc_dwc2_bus_reset(const struct device *const dev)
 	k_msleep(RESET_RECOVERY_MS);
 
 	/* Finish the port config for the appeared device */
+	dwc2_hal_set_fifo_sizes(dwc2, config->ghwcfg2, config->ghwcfg3);
+
 	/* TODO: set frame list for the ISOC/INTR xfer */
 	/* TODO: enable periodic transfer */
 
```

This also matches what is described by the datasheet.

roma-jam commented on drivers/usb/uhc/uhc_dwc2_nrf_usbhs_nrf54l.h:
```diff
@@ -0,0 +1,185 @@
+	/* Request PCLK24M using clock control driver */
+	sys_notify_init_spinwait(&data->pclk24m_cli.notify);
+	err = onoff_request(data->pclk24m_mgr, &data->pclk24m_cli);
+	if (err != 0) {
```
Probably: 


```suggestion
	if (err < 0) {
```

Because from `onoff.h`: 
```
 * @retval >=0 the observed state of the machine at the time
 * the request was processed, if successful.
 * @retval -EIO if service has recorded an error.
 * @retval -EINVAL if the parameters are invalid.
 * @retval -EAGAIN if the reference count would overflow.
```

josuah replied:
Yes, good catch, applied thank you!

jfischer-no replied:
Fixed. L:75

roma-jam commented on drivers/usb/uhc/uhc_dwc2_nrf_usbhs_nrf54l.h:
```diff
@@ -0,0 +1,185 @@
+
+	/* Release PCLK24M using clock control driver */
+	err = onoff_cancel_or_release(data->pclk24m_mgr, &data->pclk24m_cli);
+	if (err != 0) {
```
Probably: 


```suggestion
	if (err < 0) {
```

Because from `onoff.h`: 
```
 * @retval ONOFF_STATE_TO_ON if the cancellation occurred before the
 * transition completed.
 *
 * @retval ONOFF_STATE_ON if the cancellation occurred after the
 * transition completed.
 *
 * @retval -EINVAL if the parameters are invalid.
 *
 * @retval <0 other errors produced by onoff_release().
```

josuah replied:
Yes, good catch, applied thank you!

roma-jam reviewed (commented):
Hi @josuah, 

Small notes from my side. 

I tried to run circular tests (i`nit-enable-disable-enable-disable-shutdown`) and for that the logic should be a little bit changed. 

So I tried to adopt the following pattern. 
Briefly, simplified it is:  

```
pre_init:
  enable Vreg
  enable PCLK24M
  wrapper->ENABLE = CORE | PHY 
  TASKS_START

pre_enable:
  if not enabled: wrapper->ENABLE = CORE | PHY
  release D+ override

disable:
  force D+ override
  wrapper->ENABLE = 0

shutdown:
  release PCLK24M
  disable Vreg
```

Please, let me know if you have any inputs about that.


UPD: 
I have it as a commit above your changes, so you can easily see the diff and apply everything you think is worth applying. You may find it here: https://github.com/roma-jam-lab/zephyr/pull/7

The test now works with Low- and Full-speed devices, but not with High-speed, because of the error on the second enumeration. 
Apparently I missed something during enable/disable, when High-speed device is attached to the port.

UPD: I changed the quirks a bit but didn’t clean the code yet. 

The thing that fixed High speed was imlementation of missing DO_REINIT logic when tranfer XactErr | NAK | ChlHltd.


I will cleanupy additional commit for quirks and let you know when it’s ready.

josuah commented:
I think that makes a lot of sense given the PHY powers the core.

It needs to be enabled for init, but then it still needs to be enabled/disabled upon request.

I am tempted to cheat though:

```
pre_init:
  enable Vreg
  enable PCLK24M
  wrapper->ENABLE = CORE | PHY 
  TASKS_START

post_init:
  wrapper->ENABLE = 0    << only difference, we disable the core to put it on "idle"

pre_enable:
  wrapper->ENABLE = CORE | PHY     << and re-enable it back
  release D+ override

disable:
  force D+ override
  wrapper->ENABLE = 0

shutdown:
  release PCLK24M
  disable Vreg
```

This way this simplifies the logic a bit.

We must make sure that there is no register access when not in "enable" state, but that's the duty of the driver to make sure that does not happen I think?

Let's try.

roma-jam commented:
Hi @josuah, 

> It needs to be enabled for init, but then it still needs to be enabled/disabled upon request.

Yes, +1. 

I will try it with test and check that it doesn't change the stability on high amount of iterations.

I don't have anything against adding the `post_init` quirk call.

josuah commented:
There was something odd with what I proposed: if the core is effectively powered by the PHY, then `ENABLE = 0` would have been turning off DWC2 altogether, and require a reinit.

So I suspect it is just one part that is having a dependency on the PHY power, and `ENABLE = 0` does not require to re-init DWC2.

josuah commented:
I am currently facing an SoC freeze by constantly spamming enable/disable while also plugging/unplugging an USB dongle.

I am currently investigating this.

josuah commented:
Well, `wrapper->ENABLE = 0` is causing some trouble in some possibly very rare edge case.

I am tempted to just let it enabled at all time, only turning it off during "shutdown", letting enable/disabled to the minimum amount of duty they need to do, which is not much if anything at all?

josuah commented:
@roma-jam here is now a much simpler attempt:

```
pre_init:
  enable everything

shutdown:
  turn off everything
```

What do you think? I am not sure we needed to do anything for `enable()`/`disable()`, it works to just turn the IRQs on/off.

For saving power, then maybe this is the hibernation API.

roma-jam commented:
Hi @josuah, 

> I am currently facing an SoC freeze by constantly spamming enable/disable while also plugging/unplugging an USB dongle.

Hm, that strange....I don't have anything similar that fits this description.

One thing I bumped into with shell sample: I removed `default y` from Kconfig and during the build I have: `undefined reference to '__device_dts_ord_140'`. Which is clear. 


Otherwise, I do not have any problems with attaching/detaching USB device after `usbh enable` command and I also don't have any problems when I run `usbh enable` and then `usbh disable` and `usbh enable` again, while device is kept attached to the port. 

Could you please share a bit more details, how did you get this error? 


Meanwhile, I will try to add disable and enable core during `uhc_enable`/`uhc_disable` and check the state.

josuah commented:
> I removed `default y` from Kconfig and during the build I have: undefined reference to '__device_dts_ord_140'. Which is clear.

Agreed, drivers are always `default y` AFAIR.

josuah commented:
> Could you please share a bit more details, how did you get this error?

I hit "arrow up, enter" repeatedly with one hand, while the other hand plugs/unplugs the USB dongle I am testing with. This spams a lot of errors about debouncing and connection logs, then eventually faces a freeze.

If not powering off the DWC2 core during enable()/disable() (like now), I have no more bugs.

Use of regulators API in nRF54LM20 quirks has nothing to do with enabling/disabling `VBUS`, only powering the core on/off AFAIU, or maybe more precisely the PHY.

To me, it makes more sense that the PHY is not completely toggled on/off during `enable()`/`disable()`.

tmon-nordic commented:
> I am currently facing an SoC freeze by constantly spamming enable/disable while also plugging/unplugging an USB dongle.

Accessing DWC2 core register after ENABLE = 0 will hang. If ENABLE was 0 then after ENABLE = CORE | PHY, there must be TASKS_START (before core register access) otherwise accessing DWC2 core register will hang.

I have no idea what you were trying to achieve by putting ENABLE = 0 in `post_init`, but this is the reason why you have seen the hangs.

tmon-nordic commented:
> To me, it makes more sense that the PHY is not completely toggled on/off during `enable()`/`disable()`.

If you want to save power, then you really should implement hibernation. There is no shortcut as far as I know.

roma-jam commented on drivers/usb/uhc/Kconfig.dwc2:
```diff
@@ -0,0 +1,101 @@
+
+config UHC_DRIVER_HIGH_SPEED_SUPPORT_ENABLED
+	bool "UHC DWC2 High speed"
+	default y if SOC_SERIES_NRF54LX
```
Maybe? 

```suggestion
	default y if SOC_SERIES_NRF54L
```
Because https://github.com/zephyrproject-rtos/zephyr/blob/cdb58275997595faa129ca83ed0cfeb37542026b/soc/nordic/nrf54l/Kconfig#L13 ?

josuah replied:
Absolutely, this is also not the right place to introduce a generic `UHC_DRIVER_HIGH_SPEED_SUPPORT_ENABLED` config I think... It needs to be in the global UHC Kconfig like proposed here https://github.com/zephyrproject-rtos/zephyr/pull/108280

jfischer-no commented on drivers/usb/uhc/Kconfig.dwc2:
```diff
@@ -77,4 +78,13 @@ config UHC_DWC2_SET_ADDR_DELAY_MS
+	default 0
+	help
+	  VBUS ready event timeout for the USBHS variant of DWC2 found on nRF MCUs.
+	  If the VBUS is not ready and the Nordic USBHS controller is used, the udc_enable() is
```
"the udc_enable() is"
------^

uhc_enable()

josuah replied:
done, my bad

jfischer-no commented on drivers/usb/uhc/uhc_dwc2_nrf_usbhs_nrf54l.h:
```diff
@@ -0,0 +1,180 @@
+	wrapper->PHY.INPUTOVERRIDE = 0;
+	wrapper->PHY.OVERRIDEVALUES = 0;
+
+	/* Affect the regulator */
```
This comment seems misplaced.

josuah replied:
now corrected

roma-jam commented on drivers/usb/uhc/Kconfig.dwc2:
```diff
@@ -77,4 +78,13 @@ config UHC_DWC2_SET_ADDR_DELAY_MS
 
 	  The default value is set to 10 ms to be safe.
 
+config UHC_DWC2_USBHS_VBUS_READY_TIMEOUT
```
Does that make sense to the host logic? 

For the device, yes, absolutely. 
For the host...could there be potential situation, when we need to wait the VBUS ready? After suspend/hibernation probably, if we want to power down the controller....

Otherwise, I don't see that we might need this config to be user configurable.

josuah replied:
That is a good question. I would say ultimately yes:

https://docs.nordicsemi.com/r/bundle/ps_nrf54lm20a/page/regulators.html

<img width="277" height="253" alt="image" src="https://github.com/user-attachments/assets/43b68c78-0286-4724-ab5e-f2498f1b0673" />

"The core is powered by the PHY" is what I remember.

The PHY is powered from VBUS, an external signal.

So how VBUS is provided is still board-dependent to some level.

josuah replied:
Alternative is to not make it configurable and only introduce it *if* someone needs it.

tmon-nordic replied:
nRF54LM20DK does not have a way to generate VBUS on board, but at the same time the software will hang if VBUS is not these to power the PHY. Because this leads to hang, we do need some sanity check for VBUS presence.

roma-jam replied:
> Because this leads to hang, we do need some sanity check for VBUS presence.

Yes, this I can understand, thanks. 

The question was more about do we really need the configurable timeout or just sanity check it enough?

josuah replied:
I am imagining a board that needs i.e. 200 ms to stabilize VBUS as the on-board regulator has a slow startup, while another is having VBUS ready before the SoC starts (0 ms).

roma-jam reviewed (commented):
Only one question from my side regarding the necessity of having the config for VBUS ready timeout. Apart of that LGTM.   

My current tests are fine, so nothing to add ATM. 

Having only `pre_init()` and `shutdown()` seems enough.

josuah commented:
close-open for tests

tmon-nordic commented on drivers/usb/uhc/uhc_dwc2_nrf_usbhs_nrf54l.h:
```diff
@@ -0,0 +1,175 @@
+		return -ETIMEDOUT;
+	}
+
+	/* Set ID to Host and force D+ pull-up off for now */
```
I will check whether forcing D+ off is necessary - ID to Host should already prevent D+ pull-up.

josuah replied:
removing the override for D+/D- worked in the tests below

tmon-nordic commented on drivers/usb/uhc/uhc_dwc2_nrf_usbhs_nrf54l.h:
```diff
@@ -0,0 +1,175 @@
+
+	wrapper->ENABLE = 0;
+	wrapper->PHY.INPUTOVERRIDE = 0;
+	wrapper->PHY.OVERRIDEVALUES = 0;
```
```suggestion
	wrapper->ENABLE = 0;
```

What do you want to achieve here? I think that keeping `PHY.INPUTOVERRIDE` and `PHY.OVERRIDEVALUES` intact here should be fine. The ID to Host override should be there, and there is no known glitch scenario to avoid here (unlike DWC2 udc).

We may want to revisit override handling once we have runtime host/device switching, but for now just keep it simple and remove the extra writes (especially that they are not quite correct being after `ENABLE = 0`).

josuah replied:
This was to remove any override while not using the core, out of ignorance really... just to release the resources.

This still works without resetting `INPUTOVERRIDES = 0` (as seen below).

Thank you.

tmon-nordic replied:
Measurements confirmed that the pull-downs turn on after overriding ID alone (which is not surprising because it works as designed). D+ pull-up is not enabled when pull-downs are enabled so there is no need to do any more overrides.

tmon-nordic commented on drivers/usb/uhc/uhc_dwc2_nrf_usbhs_nrf54l.h:
```diff
@@ -0,0 +1,175 @@
+	/* Enable the DWC2 core */
+
+	/* Release PHY power-on reset and wait for it to start */
+	wrapper->ENABLE = USBHS_ENABLE_PHY_Msk | USBHS_ENABLE_CORE_Msk;
```
This sequence looks incorrect. The register writes can actually be lost if CONSTLAT is not enabled as there is nothing holding power to USBHS wrapper.

```suggestion
	/* Enable the DWC2 core */
	wrapper->ENABLE = USBHS_ENABLE_CORE_Msk;
	
	/* Set ID to Host */
	wrapper->PHY.INPUTOVERRIDE = USBHS_PHY_INPUTOVERRIDE_ID_Msk;
	wrapper->PHY.OVERRIDEVALUES = 0;
	
	/* Release PHY power-on reset and wait for it to start */
	wrapper->ENABLE = USBHS_ENABLE_PHY_Msk | USBHS_ENABLE_CORE_Msk;
```

Please check if the sequence that I proposed works with your development setup. I only checked it against the design, did not do measurements on hardware (I will do measurements tomorrow).

josuah replied:
It does work as tested with the board below. Thank you, I wasn't familiar enough with the PHY <-> core interaction to spot this.

josuah commented:
Testing again after the changes:

With `regulator_enable(cfg->vregusb_dev)` present, the SoC can be initialized:

```
*** Booting Zephyr OS build v4.4.0-7534-gf5d4ede646b0 ***
uart:~$ usbh init 
host: USB host initialized
[00:00:02.298,179] <dbg> uhc_dwc2.uhc_dwc2_init: DWC2 Core rev. 500b
[00:00:02.298,194] <dbg> uhc_dwc2.dwc2_core_soft_reset: DWC2 core reset done
[00:00:02.303,193] <dbg> uhc_dwc2.dwc2_core_init_host_gusbcfg: Number of Host Channels 16
[00:00:02.303,200] <dbg> uhc_dwc2.dwc2_core_init_host_gusbcfg: Highspeed UTMI+ PHY init
```

With `regulator_enable(cfg->vregusb_dev)` commented out, it prevents the access to the core correctly (error is better than freezing the SoC):

```
uart:~$ usbh init 
host: Failed to initialize -116
[00:00:27.985,676] <err> uhc_dwc2: Timed out while waiting VBUS to be ready
[00:00:27.985,684] <err> uhc_dwc2: Quirk pre init failed -116
[00:00:27.985,690] <err> uhs: Failed to init device driver
```

roma-jam reviewed (approved):
I've checked the last changes with `control` and `bulk` tests from here: https://github.com/roma-jam-lab/zephyr/pull/7 on nRF54LM20 board and they work as before. 

UPD: 
There I have tests + workflow to run them in CI on esp32s3, but when nRF quirks will be merged, I will split them and prepare a runner with nRF54 board as well. In such case, I can cherry-pick the changes and push them to my downstream repo to run the CI ob both platforms. 
This will give us some verification and certainty that at least the basic part is working.
